samples/rust/rust_inplace.rs

```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust minimal in-place sample.
+
+use kernel::prelude::*;
+
+module! {
+    type: RustInPlace,
+    name: "rust_inplace",
+    author: "Rust for Linux Contributors",
+    description: "Rust minimal in-place sample",
+    license: "GPL",
+}
+
+#[pin_data(PinnedDrop)]
+struct RustInPlace {
+    numbers: Vec<i32>,
+}
+
+impl kernel::InPlaceModule for RustInPlace {
+    type Init = impl PinInit<Self, Error>;
+    fn init(_module: &'static ThisModule) -> Result<Self::Init> {
+        pr_info!("Rust minimal sample (init)\n");
+        pr_info!("Am I built-in? {}\n", !cfg!(MODULE));
+
+        let mut numbers = Vec::new();
+        numbers.try_push(72)?;
+        numbers.try_push(108)?;
+        numbers.try_push(200)?;
+
+        Ok(try_pin_init!(Self { numbers }))
+    }
+}
+
+#[pinned_drop]
+impl PinnedDrop for RustInPlace {
+    fn drop(self: Pin<&mut Self>) {
```

Comment:
We have a few folks suggest that instead of using a drop implementation, that we should do have `deinit` or `exit` method in `kernel::Module`.

We decided against it back then, which I think was the right decision.

But I think for the pinned/static case,the ergonomics of a method in the module to clean up (which takes a `Pin<&mut self>`) are better I think.

We should perhaps consider it.

What do you all think?

Reply:
With pin-ini, we have to add `(PinnedDrop)` to `pin_data` attribute of the module, then we have to add the `#[pinned_drop]` attribute to the `PinnedDrop` impl block.

These are all magic as far as I can tell with no easy way for someone who knows about regular `Drop` impls to find this.

Reply:
I see your point, however if you have an explicit `exit` function on `Module`, then that could be called multiple times by user code.
You could make the function `unsafe`, but that feels weird to me, since then you could no longer grep for `unsafe` to check if they use `unsafe` code.
You could of course mimic the `#[pinned_drop]` macro that makes sure nobody can call the function accidentally, but then we could also just use the drop approach.

I am not sure how we could improve the situation...

Reply:
To make it not callable from other places, we can have an extra argument of some type T whose constructor is unsafe.

Note, however, that I'm not convinced yet that an `exit` function in `Module` is necessarily the way to go. I'm just saying we should consider it.

Reply:
> To make it not callable from other places, we can have an extra argument of some type T whose constructor is unsafe.

That is how `#[pinned_drop]` currently does things (it adds this parameter to the function). I think it would also be a source of confusion for users to have this visible.

Another thing to consider with the `exit` function approach is that we then need to have a blanket impl of `Drop` for all modules. Users will not be able to have a normal drop function, which I think is unfortunate, since that will also create confusion.